### PR TITLE
c2c: deflake c2c/shutdown roachtests

### LINKED ISSUE #101898: roachtest: c2c/shutdown/src/worker failed
roachtest.c2c/shutdown/src/worker [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/9699489?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/9699489?buildTab=artifacts#/c2c/shutdown/src/worker) on release-23.1 @ [ad6ce866ea3b5c5bb47ba9a0ac19b721a0c98add](https://github.com/cockroachdb/cockroach/commits/ad6ce866ea3b5c5bb47ba9a0ac19b721a0c98add):

```
test artifacts and logs in: /artifacts/c2c/shutdown/src/worker/run_1
(assertions.go:262).Fail: 
	Error Trace:	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:546
	            				github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:655
	            				github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:1046
	            				main/pkg/cmd/roachtest/monitor.go:105
	            				golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75
	            				GOROOT/src/runtime/asm_amd64.s:1594
	Error:      	Received unexpected error:
	            	expected job status succeeded, but got running
	            	(1) attached stack trace
	            	  -- stack trace:
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*replicationDriver).stopReplicationStream.func1
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:542
	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*replicationDriver).stopReplicationStream
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:529
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*replicationDriver).main
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:655
	            	  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerClusterReplicationResilience.func1.3
	            	  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:1046
	            	  | main.(*monitorImpl).Go.func1
	            	  | 	main/pkg/cmd/roachtest/monitor.go:105
	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	            	  | 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75
	            	  | runtime.goexit
	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
	            	Wraps: (2) expected job status succeeded, but got running
	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
	Test:       	c2c/shutdown/src/worker
(require.go:1264).NoError: FailNow called
(monitor.go:127).Wait: monitor failure: monitor task failed: context canceled while waiting for job to finish: context canceled
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=8</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101527 roachtest: c2c/shutdown/src/worker failed [C-test-failure O-roachtest O-robot T-disaster-recovery branch-master]
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*c2c/shutdown/src/worker.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-27178

### LINKED ISSUE #102111: roachtest: c2c/shutdown/dest/worker failed
roachtest.c2c/shutdown/dest/worker [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/9761090?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/9761090?buildTab=artifacts#/c2c/shutdown/dest/worker) on master @ [1f3419e178bdba544f74d9c9e14a4682efd18028](https://github.com/cockroachdb/cockroach/commits/1f3419e178bdba544f74d9c9e14a4682efd18028):

```
test artifacts and logs in: /artifacts/c2c/shutdown/dest/worker/run_1
(sql_runner.go:218).Scan: error scanning '&{0xc0039a3200 <nil>}': pq: system-jobs-scan: rpc error: code = Unavailable desc = connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)"
(1) pq: system-jobs-scan: rpc error: code = Unavailable desc = connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)"
Error types: (1) *pq.Error
(monitor.go:127).Wait: monitor failure: monitor task failed: context canceled while waiting for job to finish: context canceled
(test_runner.go:1087).func1: 1 dead node(s) detected
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=8</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*c2c/shutdown/dest/worker.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-27286

### PR DESCRIPTION
**Pull request by msbutler**
c2c: deflake c2c/shutdown roachtests

    This patch addresses to roachtest failure modes:
    - Prevents roachtest failure if a query fails during a node shutdown.

    - Prevents the src cluster from returning a single node topology, which could
      cause the stream ingestion job to hang if the participating src node gets
    shut down. Longer term, automatic replanning will prevent this. Specifically,
    this patch changes the kv workload to split and scatter the kv table across the
    cluster before the c2c job begins.

    Fixes #101898
    Fixes #102111

    This patch also makes it easier to reproduce c2c roachtest failures by plumbing
    a random seed to several components of the roachtest driver.

    Release note: None

    c2c: rename completeStreamIngestion to applyCutoverTime

    Release note: none

    workload: add --scatter flag to kv workload

    The user can now run `./workload init kv --scatter ....` which scatters the kv
    table across the cluster after the initial data load. This flag is best used
    with `--splits`, `--max-block-bytes`, and `--insert-count`.

    Epic: none

    Release note: none

### CODE REVIEW

File: pkg/cmd/roachtest/tests/cluster_to_cluster.go [hunk 1]
```diff
@@ -963,7 +982,10 @@ func (rrd *replResilienceDriver) waitForTargetPhase() error {
 			// Every C2C phase should last at least 30 seconds, so introduce a little
 			// bit of random waiting before node shutdown to ensure the shutdown occurs
 			// once we're settled into the target phase.
-			randomSleep := time.Duration(rand.Intn(6))
+			//
+			// Sleep for a minimum of 5 seconds to ensure the stream processors have
+			// observed the cutover signal.
+			randomSleep := time.Duration(5 + rand.Intn(6))
 			rrd.t.L().Printf("In target phase! Take a %d second power nap", randomSleep)
 			time.Sleep(randomSleep * time.Second)
```
Comment (lidorcarmel) [inline pkg/cmd/roachtest/tests/cluster_to_cluster.go:990]: I know my comment isn't really about this pr but should `waitForTargetPhase` do what it says, and just wait for the target phase without any sleep? and then the caller (node shutdown or something else) do the sleep instead?
when reading this code it's not so clear why node shutdown is mentioned in the comment.
WDYT?

File: pkg/cmd/roachtest/tests/cluster_to_cluster.go [hunk 2]
```diff
@@ -617,10 +637,9 @@ func (rd *replicationDriver) main(ctx context.Context) {
 	rd.t.Status("starting replication stream")
 	rd.metrics.initalScanStart = newMetricSnapshot(metricSnapper, timeutil.Now())
 	ingestionJobID := rd.startReplicationStream(ctx)
-
 	removeTenantRateLimiters(rd.t, rd.setup.dst.sysSQL, rd.setup.dst.name)
 
-	lv := makeLatencyVerifier("stream-ingestion", 0, 2*time.Minute, rd.t.L(), getStreamIngestionJobInfo, rd.t.Status, false)
+	lv := makeLatencyVerifier("stream-ingestion", 0, 2*time.Minute, rd.t.L(), getStreamIngestionJobInfo, rd.t.Status, true)
```
Comment (lidorcarmel) [inline pkg/cmd/roachtest/tests/cluster_to_cluster.go:642]: why do we want to ignore errors here?
don't we want the test to fail if we can't read from `crdb_internal.system_jobs`?

**Review by lidorcarmel — COMMENTED** [review]
looks good, a couple of small things..

**Review by stevendanna — APPROVED** [review]
I wonder if it would also help to explicitly scatter the ranges.

**Review by renatolabs — COMMENTED** [review]
All ignorable nits, don't mind me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-:-NV6GrjK7j21YCwiIzPH:b-h4kk6b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel and @msbutler)

___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FCAZ25xKnxPLX4gL:-NV6FCA_67luzv0Tz3pX:babitws) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L243)):*
> ```Go
> 
> func (kv replicateKV) sourceInitCmd(tenantName string, nodes option.NodeListOption) string {
> 	cmd := roachtestutil.NewCommand(`./workload init kv`)
> ```

Nit: you can chain all of these calls which arguably makes it nicer to read:

```go
return roachtestutil.NewCommand("./workload init kv").
    Flag("foo", 1).
    Flag("bar", 2).
    String()
```
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 245 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FhN8AaFoolvtRaqC:-NV6FhN97FLrUB7Dwp20:b-y3ujxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L245)):*
> ```Go
> 	cmd := roachtestutil.NewCommand(`./workload init kv`)
> 	cmd.MaybeFlag(kv.initRows > 0, "insert-count", kv.initRows)
> 	cmd.MaybeFlag(kv.initRows > 0, "max-block-bytes", kv.maxBlockBytes)
> ```

Did you want to check for `maxBlockBytes` here?
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 248 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FSsNC4AfYiqFOKA8:-NV6FSsNC4AfYiqFOKA9:b-i0d69m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L248)):*
> ```Go
> 	cmd.Flag("splits", 100)
> 	cmd.Option("scatter")
> 	cmd.Arg(fmt.Sprintf("{pgurl%s:%s}", nodes, tenantName))
> ```

Nit: `Arg` is for single argument, and it already understands `Printf`-style formatting, so you can just do `Arg("{pgurl%s:%s}", nodes, tenantName)`
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 1006 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6G0r9AmhU0-Au8w7G:-NV6G0r9AmhU0-Au8w7H:b-77s3ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L1006)):*
> ```Go
> 	// once we're fully settled into the target phase (e.g. the stream ingestion
> 	// processors have observed the cutover signal).
> 	randomSleep := time.Duration(5 + rrd.rng.Intn(6))
> ```

Nit: making `randomSleep` a `time.Duration` (i.e., number of nanoseconds), even though the actual intended meaning is for it to be a number of seconds is slightly confusing. Suggestion:

```go
dur := (5 + rrd.rng.Intn(6)) * time.Second
t.L().Printf("taking a %s power nap", dur)
time.Sleep(dur)
```

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102382)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FCAZ25xKnxPLX4gL:-NVAho_Y1HkqPVG-JW0V:b-mvxkxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L243)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: you can chain all of these calls which arguably makes it nicer to read:

```go
return roachtestutil.NewCommand("./workload init kv").
    Flag("foo", 1).
    Flag("bar", 2).
    String()
```
</blockquote></details>

TIL. Done.
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 245 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FhN8AaFoolvtRaqC:-NVAhqYE4O9EqPsHA0LY:b-iis3ls) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L245)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Did you want to check for `maxBlockBytes` here?
</blockquote></details>

Yeah I did this bc the parameter only seems relevant if i actually insert data, and bc I distrust the workload package's parameter handling logic :D
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 248 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FSsNC4AfYiqFOKA8:-NVAi8FeCSDppf7Bm4z3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L248)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: `Arg` is for single argument, and it already understands `Printf`-style formatting, so you can just do `Arg("{pgurl%s:%s}", nodes, tenantName)`
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 1006 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6G0r9AmhU0-Au8w7G:-NVAiATl8eX4yRLF5-k8:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L1006)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: making `randomSleep` a `time.Duration` (i.e., number of nanoseconds), even though the actual intended meaning is for it to be a number of seconds is slightly confusing. Suggestion:

```go
dur := (5 + rrd.rng.Intn(6)) * time.Second
t.L().Printf("taking a %s power nap", dur)
time.Sleep(dur)
```
</blockquote></details>

Done

**Comment by msbutler** [comment]
thanks for the nits @renatolabs !

**Comment by msbutler** [comment]
TFTRs!!

bors r=stevendanna

**Pull request merged**